internal/provider/resource_gitlab_runner.go

```diff
@@ -0,0 +1,257 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/xanzy/go-gitlab"
+)
+
+var _ = registerResource("gitlab_runner", func() *schema.Resource {
+	return &schema.Resource{
+		Description: `The ` + "`gitlab_runner`" + ` resource allows registering a runner, either at an instance level
+or at a group level. The runner will be registered at a group level if the token used is from a group, or at an 
+instance level if the token used is for the instance.
+
+**Upstream API**: [GitLab REST API docs](https://docs.gitlab.com/ee/api/runners.html#register-a-new-runner)`,
+
+		// Since this resource updates an in-place resource, the update method is the same as the create method
+		CreateContext: resourceGitLabRunnerCreate,
+		UpdateContext: resourceGitLabRunnerUpdate,
+		ReadContext:   resourceGitLabRunnerRead,
+		DeleteContext: resourceGitLabRunnerDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"token": {
```

Comment:
Should we call it `registration_token` explicitly? 

@armsnyder WDYT?

Reply:
I debated this myself. I decided that I'd try to stick as closely to the API as I could, which is why I changed only the second one to authentication token. That having been said, changing it would be clearer to someone who is new to the provider which token it's expecting. I could absolutely go either way!

Reply:
Yes, usually it's best to stick to the API wording.

In this case I just think that the `token` term is kinda overloaded in this case: GitLab Token, Registration Token, Authentication Token ... So I thought why not explicitly name them.

But I'm also fine with just `token` :)

Reply:
Yknow, I didn't consider that `token` was in the provider block too. I'm going to change this to registration token just so we don't confuse new users.

Reply:
Updated in last commit, just to be explicit 😄